### DISCUSSION TITLE
VT-AUX Examples UT-Pools

### DISCUSSION THREAD

**Discussion by freggala** [Open-Agriculture/AgIsoStack-plus-plus#509] [General]
For the AUX functions and inputs examples, the UT-Pool files appear to play a central role, as the code itself does little more than uploading the binary object pool from those iop files and handling some events (if I got it right).
I was wondering what's inside the iop files, how I could modify them, and adapt them to my particular needs. As I failed to open them using the AgISODDOPGenerator, I got a hint from the Open-Agriculture Telegram group to open them using a free commercial tool I wouldn't mention here for reasons. However, I can only see an empty mask and the IgISOStack++-Logo icon.

So where's all the magic that registers and parameterizes those AUX functions or inputs, assigns them icons, etc? 

How could I create my own AUX application (what tools do you use/recommend) for making the IOPs? Is there a way to do all the stuff by code rather than through a static UT-Pool file (is there an example anywhere around)?

**Comment by ad3154** [comment]
> For the AUX functions and inputs examples, the UT-Pool files appear to play a central role, as the code itself does little more than uploading the binary object pool from those iop files and handling some events (if I got it right).

Correct. The way that you tell the VT that you want to expose Aux inputs or functions is by adding either `Auxiliary Function Type 2` or `Auxiliary Input Type 2` objects to your VT object pool.

> I was wondering what's inside the iop files, how I could modify them, and adapt them to my particular needs. As I failed to open them using the AgISODDOPGenerator, I got a hint from the Open-Agriculture Telegram group to open them using a free commercial tool I wouldn't mention here for reasons. However, I can only see an empty mask and the IgISOStack++-Logo icon.

AgISODDOPGenerator is for editing TC object pools, not VT object pools. Using a third party commercial tool to edit a VT object pool is probably the best route at the moment. VT object pools in .iop format are a binary file of VT objects, so they are not really human readable or editable without a tool. The format of whats in those files are described in ISO 11783-6 Annex B and varies by object type.

The Aux examples' object pools are "minimal object pools" as described by ISO 11786-6 section J.4.1 which are used "in cases where a unit is used exclusively to provide Auxiliary Inputs". So they don't contain much.

The Aux Functions example contains:
 - A working set object (which is always required)
 - A data mask (which is also required since the working set chooses an active mask)
 - Some picture graphics (which are part of the Aux function type 2 designator)
 - Three Aux Function Type 2 Objects (Object IDs 10000, 10010, and 10020)

The Aux Inputs example contains the same required objects and:
- Some different picture graphics which are part of the designator for the Aux Input Type 2 objects
- Two Aux Input Type 2 Objects (Object IDs 3000 and 3020).

> So where's all the magic that registers and parameterizes those AUX functions or inputs, assigns them icons, etc?

The icons come from their designators, which are part of the Aux objects in the .iop file. Essentially all the child objects of an Aux object are drawn as its icon with the caveat that those objects must all fit inside a softkey sized area.

They are "registered" simply by being in the object pool.

>How could I create my own AUX application (what tools do you use/recommend) for making the IOPs? Is there a way to do all the stuff by code rather than through a static UT-Pool file (is there an example anywhere around)?

Generally you'd have to use a third party tool to create an object pool that meets your needs, then write the code using our library to interact with it. 

Currently I don't think we have a way to do it all just in code, though I think we could add that easily enough - we have functions in the stack to translate a .iop file of VT objects into C++ objects, so we'd just need a way to do the reverse. The downside of such an approach would be that it would be kind of painful to do in code since you wouldn't be able to preview it before uploading it to a VT

**Reply by ad3154** [reply]
To put it another way, if we wanted to add Aux Functions to our [seeder example](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/tree/main/examples/seeder_example) all we'd have to do is add some valid `Auxiliary Function Type 2` objects to the .iop file in there, and handle the events.
